optimum/habana/diffusers/pipelines/flux/pipeline_flux.py

```diff
@@ -0,0 +1,867 @@
+from ...schedulers import GaudiFlowMatchEulerDiscreteScheduler
+from ..pipeline_utils import GaudiDiffusionPipeline
+
+
+logger = logging.get_logger(__name__)  # pylint: disable=invalid-name
+
+
+@dataclass
+class GaudiFluxPipelineOutput(BaseOutput):
+    """
+    Output class for Stable Diffusion pipelines.
+
+    Args:
+        images (`List[PIL.Image.Image]` or `np.ndarray`)
+            List of denoised PIL images of length `batch_size` or numpy array of shape `(batch_size, height, width,
+            num_channels)`. PIL images or numpy array present the denoised images of the diffusion pipeline.
+    """
+
+    images: Union[List[PIL.Image.Image], np.ndarray]
+    throughput: float
+
+
+EXAMPLE_DOC_STRING = """
+    Examples:
+        ```py
+        >>> import torch
+        >>> from optimum.habana.diffusers import GaudiFluxPipeline
+
+        >>> pipe = GaudiFluxPipeline.from_pretrained(
+        ...    "black-forest-labs/FLUX.1-schnell",
+        ...     torch_dtype=torch.bfloat16,
+        ...     use_habana=True,
+        ...     use_hpu_graphs=True,
+        ...     gaudi_config="Habana/stable-diffusion",
+        ... )
+        >>> prompt = "A cat holding a sign that says hello world"
+        >>> # Depending on the variant being used, the pipeline call will slightly vary.
+        >>> # Refer to the pipeline documentation for more details.
+        >>> image = pipe(prompt, num_inference_steps=4, guidance_scale=0.0).images[0]
+        >>> image.save("flux.png")
+        ```
+"""
+
+
+def apply_orig_rope(xq, xk, freqs_cis):
+    xq_ = xq.float().reshape(*xq.shape[:-1], -1, 1, 2)
+    xk_ = xk.float().reshape(*xk.shape[:-1], -1, 1, 2)
+    xq_out = freqs_cis[..., 0] * xq_[..., 0] + freqs_cis[..., 1] * xq_[..., 1]
+    xk_out = freqs_cis[..., 0] * xk_[..., 0] + freqs_cis[..., 1] * xk_[..., 1]
+    return xq_out.reshape(*xq.shape).type_as(xq), xk_out.reshape(*xk.shape).type_as(xk)
+
+
+def apply_fused_rope(xq, xk, freqs_cis):
+    cos = freqs_cis[..., 0, 0].type_as(xq)
+    sin = freqs_cis[..., 1, 0].type_as(xq)
+    cos_ = torch.cat((cos, cos), dim=-1)
+    sin_ = torch.cat((sin, sin), dim=-1)
+
+    xq0 = xq[..., 0::2]
+    xq1 = xq[..., 1::2]
+    xq_ = torch.cat((xq0, xq1), dim=-1)
+
+    xk0 = xk[..., 0::2]
+    xk1 = xk[..., 1::2]
+    xk_ = torch.cat((xk0, xk1), dim=-1)
+
+    xq_out_ = FusedRoPE(xq_, cos_, sin_)
+    xk_out_ = FusedRoPE(xk_, cos_, sin_)
+
+    sh = xq_out_.shape
+    xq_out = xq_out_.view(*sh[:-1], 2, sh[-1] // 2)
+    dims = list(range(xq_out.ndimension()))
+    dims[-1], dims[-2] = dims[-2], dims[-1]
+    xq_out = xq_out.permute(*dims).contiguous().view(sh)
+
+    sh = xk_out_.shape
+    xk_out = xk_out_.view(*sh[:-1], 2, sh[-1] // 2)
+    dims = list(range(xk_out.ndimension()))
+    dims[-1], dims[-2] = dims[-2], dims[-1]
+    xk_out = xk_out.permute(*dims).contiguous().view(sh)
+
+    return xq_out, xk_out
+
+
+def apply_rope(xq, xk, freqs_cis):
+    rope_opt = os.getenv("GAUDI_FLUX_FUSED_ROPE")
```

Comment:
What's the reason to save original version of the rope algorithm (`apply_orig_rope`)? Is that only for debug purpose?

Reply:
> What's the reason to save original version of the rope algorithm (`apply_orig_rope`)? Is that only for debug purpose?

With the RT `GAUDI_FLUX_FUSED_ROPE` we can select to use RoPE (Rotary Positional Embedding) or fusedRoPE.   Some releases+devices may be more performant with original RoPE although we expect fusedRoPE to outperform generally (so it is set as a default).